sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/FluxInputStream.java

```diff
@@ -0,0 +1,242 @@
+import java.nio.ByteBuffer;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * An InputStream interface that subscribes to a Flux allows data to be read.
+ */
+public class FluxInputStream extends InputStream {
+
+    private ClientLogger logger = new ClientLogger(FluxInputStream.class);
+
+    // The data to subscribe to.
+    private Flux<ByteBuffer> data;
+
+    // Subscription to request more data from as needed
+    private Subscription subscription;
+
+    private ByteArrayInputStream buffer;
+
+    private boolean subscribed;
+    private boolean fluxComplete;
+    private boolean waitingForData;
+
+    /* The following lock and condition variable is to synchronize access between the reader and the
+        reactor thread asynchronously reading data from the Flux. If no data is available, the reader
+        acquires the lock and waits on the dataAvailable condition variable. Once data is available
+        (or an error or completion event occurs) the reactor thread acquires the lock and signals that
+        data is available. */
+    private final Lock lock;
+    private final Condition dataAvailable;
+
+    private IOException lastError;
+
+    /**
+     * Creates a new FluxInputStream
+     *
+     * @param data The data to subscribe to and read from.
+     */
+    public FluxInputStream(Flux<ByteBuffer> data) {
+        this.subscribed = false;
+        this.fluxComplete = false;
+        this.waitingForData = false;
+        this.data = data;
+        this.lock = new ReentrantLock();
+        this.dataAvailable = lock.newCondition();
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] ret = new byte[1];
+        int count = read(ret, 0, 1);
+        return count == -1 ? -1 : ret[0];
+    }
+
+    @Override
+    public int read(byte[] b, int off, int len) throws IOException {
+        validateParameters(b, off, len);
+
+        /* If len is 0, then no bytes are read and 0 is returned. */
+        if (len == 0) {
+            return 0;
+        }
+        /* Attempt to read at least one byte. If no byte is available because the stream is at end of file,
+           the value -1 is returned; otherwise, at least one byte is read and stored into b. */
+
+        /* Not subscribed? subscribe and block for data */
+        if (!subscribed) {
+            blockForData();
+        }
+        /* Now, we have subscribed. */
+        /* At this point, buffer should not be null. If it is, that indicates either an error or completion event
+           was emitted by the Flux. */
+        if (this.buffer == null) { // Only executed on first subscription.
+            if (this.lastError != null) {
+                throw logger.logThrowableAsError(this.lastError);
+            }
+            if (this.fluxComplete) {
+                return -1;
+            }
+            this.buffer = new ByteArrayInputStream(new byte[0]);
```

Comment:
This line feels spurious. Since the comment says it should only be null if there was an error or completion, and both tof those cases exit the method, we shouldn't need this line, right?

Reply:
I think I added this since I wanted buffer to be null only before the subscription. But now that I think about it, that doesnt need to be a semantic I follow since I have the subscribed boolean.

Reply:
Yea I'd rather throw IllegalState in that case so that it's consistent with the comments and with the intended behavior.

Reply:
done